README.md


# 概要

Emacs 28.0用 音楽ファイルのタグ修正 lisp 関数集です.
対応コーデックは m4a (mp4), oma (atrac3plus), flac, mp3 (ID3.2v2, ID3.2v3) そして wav です.
利用配布規定は GPL3 に準じます.


# このセットのバージョン

1.19


# 基本関数(必須)

1.  mf-tag-write.el   - タグの書き換えフロントエンド関数
2.  mf-lib-var.el     - 各ライブラリから参照されるコモン変数と関数
3.  mf-lib-mp4.el     - mp4(m4a) ライブラリ
4.  mf-lib-mp3.el     - oma(atrac3plus), mp3(ID3.2v2, ID3.2v3) ライブラリ
5.  mf-lib-flac.el    - flac ライブラリ
5.  mf-lib-wav.el     - wav ライブラリ


# ユーティリティ(サンプル)

1.  wtag.el           - mf-tag-write アルバム・エディット・インターエイス


# ドキュメント

1.  MF-TAG-WRITE.org
2.  ALIAS-TABLE.org - mf-tag-write で使える TAG の alias 表.
3.  WTAG.org
4.  CHENGES
5.  README.org もしくは README.md


# インストール

load-path の通った処に基本関数のファイルを(必要ならユーティリティ類も)置き、
init.el 等で (require 'mf-tag-write) します.

必要に応じて "wtag", "mf-lib-wav" も個別に require すれば利用できるようになります。



CHANGES
# -*- mode: org -*-
* Sat Mar  5 07:34:41 2022
** wtag.el:1.189
*** Emacs 28.91, 29.0.50 で動かくなったのを修正
    file-exists-p が変更されたようで、不正なファイル名だとエラーになってしまうので、
    拡張子によるチェックにして回避.

*** 他のファイル同様リビジョンが大きく変わっている.
    今までの変更過程を RCS file にすべてぶっこんだので.

** mf-tag-write.el:1.60
*** mf-image-type に gif, webp 等も追加(とりあえず使い道はない).
*** 他のファイル同様リビジョンが大きく変わっている.

* Mon Jan 31 14:51:12 2022
** mf-lib-flac:1.52
*** Fix: 時間用のダミータグもヘッダに含めて書き込んでいたのを修正.
    読み込み等でエラーになる場合がありました. 申し訳ありませんでした.

**** バグ版で生成してしまったデータの修復法
     mf-lib-flac.el をこの版に差替えて Emacs を再起動し、
     タグ無指定で mf-tag-write を実行すれば、ゴミを除去してヘッダを再パックします.

     : (mf-tag-write "file.flac")

     Backup を作らず、タイムスタンプを継承する場合

     : (mf-tag-write "file.flac" nil 'no-backup 'stamp-copy)

     ディレクトリごと指定できるコマンド repac-tag もつけておきます.
     指定したディレクトリ内の汚染された flac file だけを書き直しします.
     prefix 起動すると backup を作らず、タイムスタンプを継承します.

 #+BEGIN_SRC emacs-lisp
(require 'mf-tag-write)
(defun repac-tag (dir prefix)
  "DIR 下 の flac が汚染されていればクリーンアップする.
PREFIX 在りだとバックアップなし、タイムスタンプ継承になります."
  (interactive "D: \nP")
  (let ((files (directory-files dir t "\\.flac\\'")))
    (dolist (f files)
      (if (tag-pollution-p f)
          (apply #'mf-tag-write f (and prefix '(nil no-backup stamp-copy)))
        (message "It's Clean: %s" f)))))

(defun tag-pollution-p (flac)
  "flac が汚染タグなら non-nil を戻す."
  (let ((tags (mf-tag-read-alist flac nil t))
        result)
    (dolist (tag tags)
      (if (equal (car tag) mf-time-dummy)
          (setq result (cons tag result))))
    (if (= 2 (length result))
        'pollution)))

(provide 'repac-tag)
 #+END_SRC

*** Add: 変数 `mf-flac-delete-block'
    Music Center 2.5.0 で flac 取り込みをすると APPLICATION ブロックがつくようになったので
    これも削除できるようにしました.
    デリートしたいブロックがある場合、変数 `mf-flac-delete-block' にリストで指定します.

    これに伴ない PADDING 削除指定専用だった変数 `mf-flac-write-safe-pad' は廃止にしたので
    こちらで再設定してください.
    指定できるのは PADDING と APPLICATION で、
    削除したいメタブロックシンボルを指定します(以前の変数とは逆なので注意).

    : (setq mf-flac-delete-block '(PADDING APPLICATION))

*** Change: 変数 `mf-lib-flac-version' について
    Source 管理法を変更したため数値が大きく上がっています.
    あとこのファイルが元々見辛いとは思っていましたが、
    あとから見たとき追加履歴ごとに日付が無いのが致命的に判り難いと今頃気づいたので(遅すぎ)、
    卑近のもので判るものだけざっと追加しておきました.

* wtag 1.19 - Fri Oct 15 15:03:01 2021
** wtag
*** カスタマイズ変数 `wtag-force-load' のデフォルトを nil に変更
    編集モードになっていなくても D&D でジャケット変更できるようにする変数で、
    デフォが 300 だったため最初に 1度だけ編集していいのかの問い合わせが入り、
    これとは別に画像ロード側にも問い合わせがあるので、
    2重に問い合わせが出るようでとまどうので、素の設定のときそうならなくするのと
    安全性向上のための措置. (よく見ると一応メッセージは違う)
    
    関連する変数 `wtag-load-without-query' (デフォルト nil).
    これが t か keep になっていると、画像ロード時の問い合わせが抑止されるため、
    `wtag-force-load' が non-nil で問い合わせが出ても違和感が出ない.

    : ;; 自分的設定
    : (setq wtag-load-without-query 'keep)
    : (setq wtag-force-load 300)

    問い合わせ処理をひとつにまとめようともしたが、
    プログラム側からすると出す箇所が違うのでひとつにはできなかった.
    
*** Fix インデックス・バッファにもサフィクスを追加
    アルバム名をそのままバッファ名にしていたが、
    アルバム・ディレクトリが dired でも開かれていると、
    バッファ名が衝突してしまうため.(何故か開けていて今まで気づかず)

*** イメージ・バッファのモードラインにピクセル・サイズの表示
    本来右側に表示すべきだが、タイトルが長いと見切れて表示されなくなるので、
    タイトルの手前(左側)に表示されます.
    face `wtag-image-size' で色が変更できます.
    : (set-face-attribute `wtag-image-size nil :foreground "green" :weight 'bold)
      
*** ソートタグの中身を対応タグにバルーンする.
    例えば曲名の上にマウスポインタが重なると、
    対応するソートタグがあれば、そこにバルーン表示される.
    ;; Windows 以外だとどうなるのか知らない.

*** ウインドウの開く幅の調整
    インデックス・ウィンドウが曲数に合わせフィットする.
    変数 `wtag-pop-action' を nil にすると以前と同じ等分割になる.
    但しアートワークが上側になるようにしたので、
    位置関係まで同じにするなら以下のようにする.

    : (setq wtag-pop-action '(display-buffer-in-direction (window-height) (direction . above)))

** mf-lib-mp3 時間取得できないと明示的にエラーにする
   計算箇所で意図しないでエラーになりおかしな止まり方をしていた.


* wtag 1.18 - Sat Sep 18 11:12:21 2021
** wtag
*** mf-tag-read: 時間とビットレートの取得
    ダミータグ " *time" に秒数とビットレート(1/1000)をリストで得ます.
    alias シンボルは `*time'.
    通常 2要素のリストですが mp3 は 3要素目にシンボル `vbr' が, oma は `*' が入る事があり、
    それぞれ可変長、誤差在りの場合です.
    m4p は可変長でも何も付きません(現在未調査で未対応のため).

#+BEGIN_SRC emacs-lisp
(assq '*time (mf-tag-read-alias "01-Take It Easy.mp3"))
;; --> (*time " *time" 214 128 vbr)

;; assoc-default (または alist-get) なら CAR にタグ、 CDR にその値が得られる
(alist-get '*time (mf-tag-read-alias "01-Take It Easy.mp3"))
;; --> (" *time" 214 128 vbr)
(mf-sec-to-times 214) ; CADR
;; --> (0 3 34)
#+END_SRC

*** wtag-stat-view:内部のタグ表示.
    曲名の行で `C-c =' するとエイリアスタイプのリストでエコーエリアにすべてのタグが表示されます.
    アートワーク等のバイナリタグは含まれません.

** mf-lib-mp3 lyric 取得時のバグを fix.
   mf-tag-read-alist, mf-tag-read-arias から lyric にアクセスした場合
   正常に動かなかったのを修正.

   lyric から :dsc プロパティを無くし(これをタグと見なし誤動作していた)、
   そこにセットされていたデータを :cdsc プロパティにセットするようにしたので、
   もしこのプロパティを利用している人がいたらそう書換えてください.

** wav ライブラリを追加
  wav 標準の LIST info タグを読み ID3 タグにして書き戻します.
  これらは共存できるのですがどちらを見るかはアプリによって違うので、
  整合性が取れるよう 元の LIST info も ID3 に合わせて書換えされます.
  sjis の LIST が読めずに化けてしまっていた fre:ac でも
  ID3 を参照するようになり正常に表示されます.

  コンテナの配置も変更します.
  WMP で作った wav なのに エクスプローラからはタイトルが見えない等の問題回避をするため、
  コンテナの配置順序を変えます.
  これによりエクスプローラや Grooveミュージックでもタイトルが出るようになります.
  
  mf-tag-write へのライブリの事後追加が正常にできるかのテストを兼ねているので、
  今のところ標準ではインストールされないので、
  必要なら init.el 等で (require 'mf-tag-write) の後に (require 'mf-lib-wav) としてください.

#+BEGIN_SRC emacs-lisp
(require 'mf-lib-wav)
(mf-tag-read-alist "01 乙女 パスタに感動.wav")
;; --> (("TRCK" . "1") ("TCON" . "J-Pop") ("TALB" . "乙女 パスタに感動") ("TIT2" . "乙女 パスタに感動") ("TPE1" . "タンポポ") (" *ver" . "wav") (" *time" 273 1411))
#+END_SRC

** アーカイブの整理
  mf-lib-utility, tiny-ted を外しました.
  サンプルの体で入れていましたが、
  今では関数 mf-tag-read-alias, mf-tag-read-alist 等でタグを抜き出し
  lisp プリミティブな連想リスト関連の関数に渡すだけで平易にタグにアクセスできるので、
  特にサンプルも不要だと思われるのと、
  以前からグシャグシャな utility はサンプルとして如何なものかと思っていたので、
  lib-wav 追加でまたファイルが増えたのを契機に除外しました.

* wtag 1.17 - Sat Aug  8 08:06:36 2020
** wtag:wtag-regular-file-name
   <> 追加と変数化.

* wtag 1.16 - Fri May 29 10:15:49 2020
** Add: トラック・ナンバー・タグを変更するとファイル名もそれに合わせてリネームする.
   変数 `wtag-track-prefix-rename' が non-nil なら
   トラック番号を変更すると書き出すときにファイル名もそれに合わせてリネームする.
   デフォルトは t.

   ファイル名が "01-foo.mp3" のようなとき、タグのトラックを 2 にすると
   "02-foo.mp3" とリネームされる.
   "foo.mp3" のようにファイル名にトラック・プレフィクスがないならリネームはされない.

** Fix: 同じファイルを2度処理しまうことがあった.(mf-tag-write に2度渡してしまう)

mf-lib-var.el
;;; mf-lib-var-20200418.el

;; Copyright (C) 2020, 2021

;; Author:  <fubuki@frill.org>
;; Version: $Revision: 1.4 $$Name:  $
;; Keywords: multimedia

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; for `mf-tag-write' mf-lib-* variables.

;;; Code:

(require 'rx)

(defconst mf-lib-var-version "$Revision: 1.4 $$Name:  $")

(defvar mf-function-list  nil)
(defvar mf-lib-suffix-all nil)

(defvar mf-current-file nil "for buffer local variable.")
(defvar mf-current-mode nil "for buffer local variable.")
(defvar mf-current-func nil "for buffer local variable.")
(defvar mf-current-case nil "for buffer local variable.")

(defcustom mf-type-dummy " *ver"
  "Tag type を保存する擬似タグ. 行頭がブランクならヘッダ生成時に無視される."
  :type  'string
  :group 'music-file)

(defvar mf-type-dummy-symbol '*type)

(defcustom mf-time-dummy " *time"
  "時間関係を保存する擬似タグ. 行頭がブランクならヘッダ生成時に無視される."
  :type  'string
  :group 'music-file)

(defvar mf-time-dummy-symbol '*time)

(defcustom mf-geob-image "OMG_TDFCA"
  "oma data ジャケット格納 DSC タグ."
  :type 'string
  :group 'music-file)


(defconst mf-omg-tags '("USR_L2TMDDA" "OMG_TRLDA" "OMG_TIT2S" "OMG_ATPE1" "OMG_ATP1S"
                        "OMG_ASGTM"   "OMG_ALBMS" "OMG_AGENR" "OMG_TRACK" "OMG_TPE1S"
                        ;; for LAME
                        "ALBUMSORT" "ALBUMARTISTSORT" "TITLESORT" "ARTISTSORT"))

;; :tag "----" :mean "com.apple.iTunes"
(defconst mf-itunes-tags '("iTunSMPB" "Encoding Params" "iTunNORM" "iTunes_CDDB_IDs"
                           "UFIDhttp://www.cddb.com/id3/taginfo1.html"))

;; 予想で充てたシンボルであり公式ではない.
(defconst mf-ilst-data-type
  '((0 . binary) (1 . string) (13 . jpeg) (14 . png) (21 . number)) "ilst data type.")

(defun mf-re-suffix (lst)
  "拡張子シンボルのリスト LST にマッチする正規表現文字列を生成."
  (let ((seq (mapcar #'symbol-name lst)))
    (eval (list 'rx "." (cons 'or seq) 'string-end))))

;; ;; 拡張子マッチに限ればこれでもできる.
;; (defmacro mf-re-suffix* (lst)
;;   (concat "\\.\\("
;;           (mapconcat #'concat (mapcar #'symbol-name (eval lst)) "\\|")
;;           "\\)\\'"))

(defun mf-first   (a) (nth 0 a))
(defun mf-second  (a) (nth 1 a))
(defun mf-third   (a) (nth 2 a))
(defun mf-fourth  (a) (nth 3 a))
(defun mf-fifth   (a) (nth 4 a))
(defun mf-sixth   (a) (nth 5 a))
(defun mf-seventh (a) (nth 6 a))
(defun mf-eighth  (a) (nth 7 a))
(defun mf-ninth   (a) (nth 8 a))
(defun mf-tenth   (a) (nth 9 a))
(defun mf-pair-p  (a) (and (consp a) (cdr a) (atom (cdr a))))

;; 経過が不要なら M-x calculator 数値 H 等でお手軽に判る.
(defun octput (val)
  "数値を 8進数にして表示."
  (format "%o" val))

(defun hexput (val)
  "数値の 16進表記を得る."
  (interactive "NVal: ")
  (format "%x" val))

(defun mp4-tag-type (tag)
  "TAG の ilst 管理番号を返す.
今調べるのが面倒なので `mf-list-convert' から 数値タグ等を指定しない前提の簡易版."
  (cond
   ((member tag '("disk" "trkn"))
    (car (rassq 'binary mf-ilst-data-type)))
   (t
    (car (rassq 'string mf-ilst-data-type)))))

(defcustom mf-no-one-patch nil
  "NON-NIL なら A30 前夜の Walkman を騙すパッチを充てない."
  :type  'boolean
  :group 'music-file)

(defcustom mf-no-mc-delete nil
  "NON-NIL なら MusicCenter が作る互換用のタグブロックを削除しない."
  :type  'boolean
  :group 'music-file)

(defvar mf-list-pack-partition
  '(((ll longlong)      mf-buffer-read-longlong-word 8)
    ((l  long longword) mf-buffer-read-long-word     4)
    ((t  twenty-four)   mf-buffer-read-3-bytes       3)
    ((w  word s short)  mf-buffer-read-word          2)
    ((b  byte c char)   mf-char-after                1)))

(defun mf-buffer-substring (start end)
  (ignore-errors (buffer-substring start end)))

(defun mf-char-after (&optional pos opt)
  (let ((pos (or pos (point))))
    (char-after pos)))

(defun mf-buffer-read-word (&optional pos opt)
  "POS から word 長を返す.
POS を省略するとカレント point になる."
  (let ((pos (or pos (point)))
        high low)
    (setq high (char-after pos)
          low  (char-after (+ pos 1)))
    (if (null (and high low))
        nil
      (+ (* high 256) low))))

(defun mf-buffer-read-long-word (&optional pos opt)
  "POS から 4バイト読んで整数として返す. POS が範囲外なら NIL を返す."
  (let* ((pos (or pos (point)))
         (high (mf-buffer-read-word pos opt))
         (low  (mf-buffer-read-word (+ 2 pos) opt)))
    (if (null (and high low))
        nil
      (+ (* high 65536) low))))

(defun mf-buffer-read-longlong-word (&optional pos wlst)
  "POS から 8バイト読んで整数として返す. POS が範囲外なら NIL を返す.
WLST が non-nil なら 64bit を 16bit ごとに分割したリストにした形式で戻す."
  (let* ((pos (or pos (point)))
         (high (mf-buffer-read-long-word pos))
         (low  (mf-buffer-read-long-word (+ 4 pos))))
    (if (null (and high low))
        nil
      (if wlst
          (list (lsh high -16 ) (logand high 65536)
                (lsh low -16 ) (logand low 65536))
        (+ (* high (expt 2 32)) low)))))

(defun mf-buffer-read-3-bytes (&optional pos opt)
  (let (a b c)
    (or pos (setq pos (point)))
    (setq a (char-after pos)
          b (char-after (+ pos 1))
          c (char-after (+ pos 2)))
    (if (null (and a b c))
        nil
      (+ (* a 65536) (* b 256) c))))

(defun mf-buffer-read-word-le (&optional pos opt)
  "POS から word 長を little endian で返す.
POS を省略するとカレント point になる."
  (let ((pos (or pos (point)))
        high low)
    (setq high (char-after pos)
          low  (char-after (+ pos 1)))
    (if (null (and high low))
        nil
      (+ (* low 256) high))))

(defun mf-buffer-read-long-word-le (&optional pos opt)
  "POS から 4バイトを little endian として読んで整数として返す.
POS が範囲外なら NIL を返す."
  (let (high low a b c d)
    (or pos (setq pos (point)))
    (setq a (char-after pos)
          b (char-after (+ 1 pos))
          c (char-after (+ 2 pos))
          d (char-after (+ 3 pos)))
    (if (null (and a b c d))
        nil
      (setq high (+ (* b 256) a)
            low  (+ (* d 256) c))
      (+ (* low 65536) high))))

(defun mf-buffer-read-word-le-fd ()
  "point から word 長を little endian で返し、その分 point を進める."
    (prog1
        (mf-buffer-read-word-le)
      (forward-char 2)))

(defun mf-buffer-read-long-word-le-fd ()
  "point から long word 長を little endian で返し、その分 point を進める."
  (prog1
      (mf-buffer-read-long-word-le)
    (forward-char 4)))

(defun mf-3-byte-char (int)
  "INT を char char char の 24ビットで構成されたバイトの並びにする."
  (encode-coding-string
   (string (logand (lsh int -16) 255) (logand (lsh int -8) 255) (logand int 255))
   'iso-8859-1))

(defun mf-long-word (value)
  "VALUE をバイト分解し  4 bytes 文字列にする."
  (encode-coding-string
   (string
    (lsh value -24)
    (logand (lsh value -16) 255)
    (logand (lsh value  -8) 255)
    (logand value           255))
   'iso-8859-1))

(defun mf-list-pack (point partition wlst)
  "POINT からバッファの内容を PARTITION に従いバイトレベルで小分けしてリストにパック.
指定可能な長さは 64bit longlongword までで 各値は符合無しの整数になる.
数値変換せずバイトの羅列としてそのまま切り出すときは\
整数を指定するとその長さだけ切り出される.
PARTITION は `mf-list-pack-partition' の 各 car で定義されたシンボルをリストで羅列する.
Example. \(mf-list-pack point '(l l w l c))
WLST が non-nil なら 64bit 値  16bit ごとに分割したリストにした形式で戻す."
  (let (fun result)
    (dolist (a partition (reverse result))
      (setq fun (assoc-default a mf-list-pack-partition #'(lambda (a b) (memq b a))))
      (cond
       (fun
        (setq result (cons (funcall (car fun) point wlst) result)
              point (+ point (cadr fun))))
       ((numberp a)
        (setq result (cons (buffer-substring point (+ point a)) result)
              point (+ point a)))
       (t
        (error "Illegal %s" a))))))

(defun mf-disbits (val partition &optional width)
  "VAL をビット長 WIDTH として PARTITION に従いビット分解しリストにして戻す.
PARTITION はビット数をリストで羅列する.
WIDTH を省略すると PARTITION の合計補正し自動的に計算する.
Example: \(mf-disbits val '(20 3 5 36))"
  (let ((width
         (or width
             (let ((w (apply #'+ partition)))
               (if (zerop (% w 8))
                   w
                 (+ (- 8 (% w 8)) w)))))
        result tmp)
    (dolist (elt partition (reverse result))
      (setq tmp (lsh val (* (- width elt) -1))
            width (- width elt))
      (setq result (cons (logand tmp (1- (expt 2 elt))) result)))))

(defun mf-chop (str)
  (save-match-data
    (string-match "\\([^\0]+\\)\0*\\'" str)
    (or (match-string 1 str) "")))

;; `mf-add-longlong' のために複数の引数に対応.
(defun mf-add-word (&rest args)
  "word (16bit) 加算器.
整数 ARGS を加算し 16bit で結果の CAR に返す.
桁溢れがあれば CDR (Carry Flag) が NON-NIL になる.
ちなみに変数 result が 16ビット幅以上でない場合正しく動作しない.
BUG: 負の引数は考慮されていない."
  (let* ((result (apply '+ args))
         (carry (if (not (zerop (logand (lsh result -16) 65535))) 'carry)))
    (cons (logand result 65535) carry)))

(defun mf-write-file (file no-backup)
  "buffer の中身を丸々 FILE に書き出す.
NO-BACKUP が non-nil ならバックアップを作らない.
once ならバックアップがあればバックアップしない."
  (let ((backup (make-backup-file-name file)))
    (cond
     ((and (eq no-backup 'once) (not (file-exists-p backup)))
      (rename-file file backup))
     ((and (not (stringp no-backup)) (null no-backup))
      (when (file-exists-p backup)
        (delete-file backup 'trash))
      (rename-file file backup)))
    (write-region (point-min) (point-max) file)))

(defun mf-sec-to-times (sec)
  "整数秒 SEC を時間リスト \(h m s) に変換."
  (mapcar #'floor (list (/ sec 3600) (/ (mod sec 3600) 60) (mod sec 60))))

(defvar mf-files '("mf-tag-write" "mf-lib-mp3" "mf-lib-mp4"
                   "mf-lib-flac" "mf-lib-utility" "mf-lib-var"
                   "wtag" "tiny-ted"))
(defvar mf-local (locate-user-emacs-file "local"))
;; user-emacs-directory

(defun mf-snap (tag)
  "for wtag release tag set."
  (interactive "sTag: ")
  (let* ((files mf-files)
         (local mf-local)
         (rcs (list "rcs" (format "-n%s:" tag)))
         (co  (list "co"  (format "-r%s"  tag))))
    (cd local)
    (dolist (com (list rcs co))
      (shell-command (mapconcat #'identity (append com files) " ")))))

(provide 'mf-lib-var)
;;; mf-lib-var-20200418.el ends here
